proposals/multi-field-value-classes.md

```diff
@@ -0,0 +1,691 @@
+    }
+}
+```
+
+Using `Password` class instead of raw `String` helps you not to pass invalid password when validated one is expected and it helps not to pass it accidentally as `Username`. There is no overhead because password is represented as raw `String` in the compiled code and the runtime.
+
+### Multi-field value classes
+
+However, Inline classes are very limited: it may be useful to escape creating wrappers because of performance issues for classes with multiple fields too. Inline classes generalization is named [Value classes](https://github.com/Kotlin/KEEP/blob/master/notes/value-classes.md) (shallow immutable classes without identity).
+
+As well, as for Inline classes, one of the main aims of MFVC is the compiler optimizations that are possible due to the lack of fixed identity. Unfortunately, there are several limitations for optimization for JVM target described below in the [Boxing](#boxing) section.
+
+Multi-field value classes are the next part of the Value classes feature after Inline classes.
+
+### Value classes
+
+Value classes are effective shallow-immutable data classes without identity, copy functions and component functions.
+
+Shallow immutability is chosen to make it possible to implement internally mutable classes with immutable public API such as `Lazy<T>`. However, there are domains where deep immutability is required, but allowing such requirement is currently outside the scope of the document.
+
+Value classes do not have `copy` methods due to their inconvenience for usage to mutate:
+* They are verbose and hard to understand when combined with constructions such as `if`s, loops.
+* They are unsuitable for nested structures: `a = a.copy(b = a.b.copy(c = f(a.b.c + 1)))` instead of `a.b.c = f(a.b.c)` for mutable classes.
+* They cannot be used for operators: `a = a.copy(b = a.b + 2)` instead of `a.b += 2` for mutable classes.
+
+Not yet implemented lens syntax is going to be used for convenient mutating mutable fields/variables of immutable object types. Other applications of the shallow-copying function `copy` do not exist for shallow-immutable value classes.
+
+Value classes do not have component functions as they are used for positional destructuring which is considered harmful because of exposing of the internal state and order of the fields. It will be replaced with not yet implemented named destructuring syntax.
+
+Customizing `equals` is not allowed until “typed equals” feature is released.
+
+### Further development: `VArray`s and reification
+
+One of other important steps of Value classes feature are Value arrays (`VArray`s) and reified classes and reified not-inline functions. They significantly extend applicability and optimizations of Value classes, **but they require MFVC to be already implemented** for that. Their importance is caused by frequent usage of containers that store and use stored values: they are currently handle value classes as boxed.
```

Comment:
I mean that `impact(mfvc + varrays) > impact(mfvc) + impact(varrays)` because it also includes `VArray<MFVC>`s. I don't know how to formulate it properly.